src/Compilers/Core/Portable/SourceGeneration/Nodes/BatchNode.cs

```diff
@@ -82,6 +70,45 @@ public NodeStateTable<ImmutableArray<TInput>> UpdateStateTable(DriverStateTable.
             return newTable.ToImmutableAndFree();
         }
 
+        private (ImmutableArray<TInput>, ImmutableArray<(IncrementalGeneratorRunStep InputStep, int OutputIndex)>) GetValuesAndInputs(
+            NodeStateTable<TInput> sourceTable,
+            NodeStateTable<ImmutableArray<TInput>> previousTable,
+            NodeStateTable<ImmutableArray<TInput>>.Builder newTable)
+        {
+            var sourceValuesBuilder = s_builderPool.DequeuePooledItem();
+            try
+            {
+                var sourceInputsBuilder = newTable.TrackIncrementalSteps ? ArrayBuilder<(IncrementalGeneratorRunStep InputStep, int OutputIndex)>.GetInstance() : null;
+
+                foreach (var entry in sourceTable)
+                {
+                    // At this point, we can remove any 'Removed' items and ensure they're not in our list of states.
+                    if (entry.State != EntryState.Removed)
+                        sourceValuesBuilder.Add(entry.Item);
+
+                    // However, regardless of if the entry was removed or not, we still keep track of its step information
+                    // so we can accurately report how long it took and what actually happened (for testing validation).
+                    sourceInputsBuilder?.Add((entry.Step!, entry.OutputIndex));
+                }
+
+                var sourceInputs = newTable.TrackIncrementalSteps ? sourceInputsBuilder!.ToImmutableAndFree() : default;
+
+                // If we're producing the exact same values as the last time, then just point at the prior array
+                // wholesale to avoid a costly copy.
+                if (previousTable.Count == 1 &&
+                    sourceValuesBuilder.Builder.SequenceEqual(previousTable.Single().item, EqualityComparer<TInput>.Default))
+                {
+                    return (previousTable.Single().item, sourceInputs);
```

Comment:
Do we need to call `Single()` twice? Or can we do `previousTable is [var item]` on the first line?

Reply:
fixed to only compute once.  note, this is not Linq's .Single method, this is an actual helper method on NodeStateTable that goes and pokes into a 2d-array to get the value in question. :)